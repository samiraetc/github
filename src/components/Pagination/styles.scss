.pagination {
  display: flex;
  list-style-type: none;

  .pagination-item {
    padding: 0 12px;
    height: 32px;
    text-align: center;
    margin: auto 4px;
    color: $github-secondary-color;
    display: flex;
    box-sizing: border-box;
    align-items: center;
    letter-spacing: 0.01071em;
    border-radius: 16px;
    line-height: 1.43;
    font-size: $github-font-size-16;
    min-width: 32px;

    &.dots:hover {
      background-color: transparent;
      cursor: default;
    }
    &:hover {
      background-color: rgba(0, 0, 0, 0.04);
      cursor: pointer;
    }

    &.selected {
      background-color: $github-primary-color;
    }

    .arrow {
      &::before {
        position: relative;
        content: '';
        display: inline-block;
        width: 0.4em;
        height: 0.4em;
        border-right: 0.12em solid $github-secondary-color;
        border-top: 0.12em solid $github-secondary-color;
      }

      &.left {
        transform: rotate(-135deg) translate(-50%);
      }

      &.right {
        transform: rotate(45deg);
      }
    }

    &.disabled {
      pointer-events: none;

      .arrow::before {
        border-right: 0.12em solid $github-secondary-color;
        border-top: 0.12em solid $github-secondary-color;
      }

      &:hover {
        background-color: transparent;
        cursor: default;
      }
    }
  }
}
