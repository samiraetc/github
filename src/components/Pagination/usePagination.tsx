import { useMemo } from 'react';

export const DOTS = '...';

const range = (start: number, end: number) => {
  const length = end - start + 1;
  return Array.from({ length }, (_, idx) => idx + start);
};

export type usePaginationProps = {
  totalCount: number;
  pageSize: number;
  siblingCount: number;
  currentPage: number;
};

export const usePagination = ({
  totalCount = 1,
  pageSize = 6,
  siblingCount = 1,
  currentPage = 1,
}: usePaginationProps) => {
  const paginationRange = useMemo(() => {
    const totalPageCount = Math.ceil(totalCount / pageSize);
    const totalPageNumbers = siblingCount + 5;
    const leftSiblingIndex = Math.max(currentPage - siblingCount, 1);
    const rightSiblingIndex = Math.min(currentPage + siblingCount, totalPageCount);
    const shouldShowLeftDots = leftSiblingIndex > 2;
    const shouldShowRightDots = rightSiblingIndex < totalPageCount - 2;
    const firstPageIndex = 1;
    const lastPageIndex = totalPageCount;
    const ItemCount = 3 + 2 * siblingCount;

    if (totalPageNumbers >= totalPageCount) {
      return range(1, totalPageCount);
    }

    const shouldRenderDots = new Map([
      [!shouldShowLeftDots && shouldShowRightDots, [...range(1, ItemCount), DOTS, totalPageCount]],
      [
        shouldShowLeftDots && !shouldShowRightDots,
        [firstPageIndex, DOTS, ...range(totalPageCount - ItemCount + 1, totalPageCount)],
      ],
      [
        shouldShowLeftDots && shouldShowRightDots,
        [firstPageIndex, DOTS, ...range(leftSiblingIndex, rightSiblingIndex), DOTS, lastPageIndex],
      ],
    ]);

    return shouldRenderDots.get(true);
  }, [totalCount, pageSize, siblingCount, currentPage]);

  return paginationRange;
};
