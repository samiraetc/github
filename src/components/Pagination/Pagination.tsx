import React from 'react';
import classnames from 'classnames';
import { usePagination, DOTS } from './usePagination';

export const COMPONENT_ID = 'pagination';

export type PaginationProps = {
  onPageChange: (value: any) => unknown;
  totalCount: number;
  siblingCount?: number;
  currentPage: number;
  pageSize: number;
  className: string;
};
const Pagination = ({
  onPageChange,
  totalCount = 2,
  siblingCount = 1,
  currentPage = 1,
  pageSize = 0,
  className = '',
}: PaginationProps) => {
  const paginationRange =
    usePagination({
      currentPage,
      totalCount,
      siblingCount,
      pageSize,
    }) || [];

  const lastPage = paginationRange[paginationRange.length - 1];

  if (currentPage === 0 || paginationRange.length < 2) {
    return null;
  }

  const onNext = () => {
    onPageChange(currentPage + 1);
  };

  const onPrevious = () => {
    onPageChange(currentPage - 1);
  };

  return (
    <ul className={`${COMPONENT_ID} ${className}`} id={COMPONENT_ID}>
      <li
        className={classnames('pagination-item', {
          disabled: currentPage === 1,
        })}
        onClick={onPrevious}
      >
        <div className="arrow left" />
      </li>
      {paginationRange.map((pageNumber: any, index: number) => {
        if (pageNumber === DOTS) {
          return (
            <li key={index} className="pagination-item dots">
              &#8230;
            </li>
          );
        }

        return (
          <li
            key={index}
            className={classnames('pagination-item', {
              selected: pageNumber === currentPage,
            })}
            onClick={() => onPageChange(pageNumber)}
          >
            {pageNumber}
          </li>
        );
      })}
      <li
        className={classnames('pagination-item', {
          disabled: currentPage === lastPage,
        })}
        onClick={onNext}
      >
        <div className="arrow right" />
      </li>
    </ul>
  );
};

export default Pagination;
