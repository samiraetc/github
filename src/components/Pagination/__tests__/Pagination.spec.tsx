import React from 'react';
import { configure, fireEvent, render } from '@testing-library/react';

import Pagination, { COMPONENT_ID } from '../Pagination';

describe('Pagination', () => {
  const onClick = jest.fn();
  test('should render Pagination', () => {
    configure({
      testIdAttribute: 'id',
    });

    const { asFragment, getByTestId } = render(
      <Pagination
        className="pagination-bar"
        currentPage={1}
        totalCount={10}
        pageSize={5}
        onPageChange={() => onClick()}
      />,
    );
    expect(getByTestId(COMPONENT_ID)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render Pagination and click on second page', () => {
    const { asFragment, getByTestId } = render(
      <Pagination
        className="pagination-bar"
        currentPage={1}
        totalCount={10}
        pageSize={5}
        onPageChange={() => onClick(2)}
      />,
    );

    configure({
      testIdAttribute: 'class',
    });
    fireEvent.click(getByTestId('arrow right'));
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render Pagination and click to back to first page', () => {
    const { asFragment, getByTestId } = render(
      <Pagination
        className="pagination-bar"
        currentPage={2}
        totalCount={10}
        pageSize={5}
        onPageChange={() => onClick(2)}
      />,
    );

    configure({
      testIdAttribute: 'class',
    });
    fireEvent.click(getByTestId('arrow left'));
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render Pagination when has dots ...', () => {
    configure({
      testIdAttribute: 'id',
    });

    const { asFragment, getByTestId, getByText } = render(
      <Pagination
        className="pagination-bar"
        currentPage={1}
        totalCount={20}
        pageSize={2}
        onPageChange={() => onClick(2)}
      />,
    );
    expect(getByTestId(COMPONENT_ID)).toBeInTheDocument();
    fireEvent.click(getByText('2'));
    expect(asFragment()).toMatchSnapshot();
  });
});
