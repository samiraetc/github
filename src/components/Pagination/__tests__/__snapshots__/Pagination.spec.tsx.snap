// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should render Pagination 1`] = `
<DocumentFragment>
  <ul
    class="pagination pagination-bar"
    id="pagination"
  >
    <li
      class="pagination-item disabled"
    >
      <div
        class="arrow left"
      />
    </li>
    <li
      class="pagination-item selected"
    >
      1
    </li>
    <li
      class="pagination-item"
    >
      2
    </li>
    <li
      class="pagination-item"
    >
      <div
        class="arrow right"
      />
    </li>
  </ul>
</DocumentFragment>
`;

exports[`Pagination should render Pagination and click on second page 1`] = `
<DocumentFragment>
  <ul
    class="pagination pagination-bar"
    id="pagination"
  >
    <li
      class="pagination-item disabled"
    >
      <div
        class="arrow left"
      />
    </li>
    <li
      class="pagination-item selected"
    >
      1
    </li>
    <li
      class="pagination-item"
    >
      2
    </li>
    <li
      class="pagination-item"
    >
      <div
        class="arrow right"
      />
    </li>
  </ul>
</DocumentFragment>
`;

exports[`Pagination should render Pagination and click to back to first page 1`] = `
<DocumentFragment>
  <ul
    class="pagination pagination-bar"
    id="pagination"
  >
    <li
      class="pagination-item"
    >
      <div
        class="arrow left"
      />
    </li>
    <li
      class="pagination-item"
    >
      1
    </li>
    <li
      class="pagination-item selected"
    >
      2
    </li>
    <li
      class="pagination-item disabled"
    >
      <div
        class="arrow right"
      />
    </li>
  </ul>
</DocumentFragment>
`;

exports[`Pagination should render Pagination when has dots ... 1`] = `
<DocumentFragment>
  <ul
    class="pagination pagination-bar"
    id="pagination"
  >
    <li
      class="pagination-item disabled"
    >
      <div
        class="arrow left"
      />
    </li>
    <li
      class="pagination-item selected"
    >
      1
    </li>
    <li
      class="pagination-item"
    >
      2
    </li>
    <li
      class="pagination-item"
    >
      3
    </li>
    <li
      class="pagination-item"
    >
      4
    </li>
    <li
      class="pagination-item"
    >
      5
    </li>
    <li
      class="pagination-item dots"
    >
      â€¦
    </li>
    <li
      class="pagination-item"
    >
      10
    </li>
    <li
      class="pagination-item"
    >
      <div
        class="arrow right"
      />
    </li>
  </ul>
</DocumentFragment>
`;
